Devops 157 support developers

deepfire opened the pull request:
- dev-env key generation
- DNS provisioning in the `aws.iohkdev.io` domain

domenkozar commented on iohk/iohk-ops.hs:
```diff
@@ -289,4 +289,44 @@ runTemplate o@Options{..} Template{..} args = do
+runNew _ _ _ = error "impossible"
+
+-- | Use 'cardano-keygen' to create keys for a develoment cluster.
+generateDevKeys :: Options -> ConfigurationKey -> Turtle.FilePath -> IO ()
```
This should be named `generateGenesis` as a general case, then used for development clusters as needed.

deepfire replied:
The genesis is generated by `cardano-node` during startup, so it's really just generation of keys compatible with a provided configuration key -- to the best of my understanding.

cc @gromakovsky

domenkozar replied:
Why is there parameter called `--genesis-out-dir`?

In that case we should give it a bit descriptive name, like `generateStakeKeys`.

gromakovsky replied:
Probably because @pva701 is bad at naming things :)
In master we also have `dump-genesis-data --path PATH` command.
I can rename `generate-keys-by-spec` to `dump-genesis-secrets` (not keys, because it also dumps AVVM seeds) and its path argument to `--path`. I'd like to do it in master, because it's not something strictly necessary for 1.0.

domenkozar replied:
I'd still call this function `generateStakeKeys` as that's what it does, the fact that it's developer environment comes from it's usage, not definition.

domenkozar commented on modules/development.nix:
```diff
@@ -7,8 +7,8 @@ with import ./../lib.nix;
       enableEkgWeb                 = true;
-      dnsHostname                  = null;
-      dnsDomainname                = mkForce null;
+      dnsHostname                  = "${name}-${config.deployment.name}";
```
`name` has the same value as `config.deployment.name`, did you verify the resulting URL?

deepfire replied:
@domenkozar, are you sure?

I'm reading https://nixos.org/nixops/manual/#ssec-relnotes-1.3 differently.

Also: https://github.com/NixOS/nixops/blob/master/nix/eval-machine-info.nix#L73

deepfire replied:
As to the DNS provisioning -- I've tested it, indeed.

domenkozar replied:
You're right, `name` is machine name.

deepfire commented:
@domenkozar, CI is green now.

domenkozar commented:
@deepfire is it complete so I can re-review? Note that it has git conflicts.

deepfire commented:
@domenkozar, fixed the `cardano-sl-src.json` bumping conflict.

Answering your question -- I believe I have addressed all of your review comments, yes.

domenkozar commented on iohk/NixOps.hs:
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
+iohkNixopsPath :: FilePath
+iohkNixopsPath =
+  let defaultNix = "default.nix"
+      storePath  = Sys.unsafePerformIO $ inprocs "nix-build" ["-A", "nixops", format fp defaultNix] $
```
Why not just make it an `IO` function?

deepfire replied:
The intent here is to calculate it just once -- an IO action would be recomputed repeatedly.

domenkozar commented on iohk/NixOps.hs:
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
-nixops  :: Options -> NixopsConfig -> NixopsCmd -> [Arg] -> IO ()
-nixops' = nixops'' cmd'
-nixops  = nixops'' cmd
+nixops'  :: Options -> NixopsConfig -> NixopsCmd -> [Arg] -> IO (ExitCode, Text)
```
It's really really hard to follow functions like this. There are 5 functions used in below two lines with almost indentical names. Why not give them appropriate names?

deepfire replied:
I believe I didn't change this code in this PR.  Can we address this later?

deepfire replied:
Also, do you have any suggestions?

domenkozar replied:
Sure

domenkozar commented on iohk/NixOps.hs:
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
                    (trace (T.unpack $ format ("INFO: using "%fp%" expression for its definition of 'nixops'") defaultNix) empty)
-      nixopsPath = Path.fromText $ T.strip storePath <> "/bin/nixops"
-  in trace (T.unpack $ format ("INFO: nixops is "%fp) nixopsPath) nixopsPath
+      opsPath    = Path.fromText $ T.strip storePath <> "/bin/nixops"
```
We have `nixops` and `iohk-ops`, so `ops` is an overloaded term, why make the name binding ambigious?

deepfire replied:
This is a local variable, and the meaning is very clear, IMO -- its usage is really spread across three lines.   But ok.
